src/hypothesis/_settings.py

```diff
@@ -521,18 +490,44 @@ def _validate_timeout(n):
     future_default=False,
 )
 
+
+def _validate_database(db):
+    from hypothesis.database import ExampleDatabase
+    if db is None or db is not_set or isinstance(db, ExampleDatabase):
+        return db
+    if db not in _db_cache:
+        _db_cache[db] = ExampleDatabase(db)
+    return _db_cache[db]
+
+
 settings.define_setting(
-    'database_file',
-    default=lambda: (
-        os.getenv('HYPOTHESIS_DATABASE_FILE') or
-        os.path.join(hypothesis_home_dir(), 'examples')
-    ),
+    'database',
+    default=lambda: _validate_database(infer),
     show_default=False,
     description="""
-    An instance of hypothesis.database.ExampleDatabase that will be
+An instance of hypothesis.database.ExampleDatabase that will be
 used to save examples to and load previous examples from. May be None
-in which case no storage will be used.
-"""
+in which case no storage will be used, `:memory:` for an in-memory
+database, or any path for a directory-based example database.
+""",
+    validator=_validate_database,
+)
+
+settings.define_setting(
+    'database_file',
+    default=not_set,
+    show_default=False,
+    description="""
+The directory location to save and load previously tried examples;
```

Comment:
Unfortunately subtly wrong due to the complication of the old sqlite format. It could be either a directory or an ordinary file depending on the file name due to the fallback logic.